### PR TITLE
Flushes append vec mmaps only if dirty

### PR DESCRIPTION
**Pull request by brooksprumo**
#### Problem

We flush append vecs *a lot*. Too much, some might say[^1]. 

Some observations: the common case is append-vecs are write-once, yet we call `flush()` on every append vec for every snapshot. The default is every 100 slots. So we are doing a lot of unnecessary work asking the OS/kernel to re-flush something that doesn't need to be flushed again.

[^1]: @alessandrod 

#### Summary of Changes

Track if an AppendVec's internal mmap is dirty or not. If it has been written to, then it is dirty. If it's dirty, then we flush. Otherwise, no need.

#### Results

On the left half of the graph is my node running regular master. On the right half I restarted the node with this PR.

Before, flushing took about 1 second. Afterwards, flushing takes about 20 milliseconds.

Screenshot 2024-06-10 at 7 27 12 PM

In addition to these speedups, we should also see improvements to the system as-a-whole, due to stress reduction of not asking the kernel to flush 432,000 files every 100 slots.

### CODE REVIEW

File: accounts-db/src/append_vec.rs [hunk 1]
```diff
@@ -212,7 +212,19 @@ struct AccountOffsets {
 #[derive(Debug)]
 enum AppendVecFileBacking {
     /// A file-backed block of memory that is used to store the data for each appended item.
-    MmapOnly(MmapMut),
+    MmapOnly(MmapOnly),
```
Comment (brooksprumo) [inline accounts-db/src/append_vec.rs:215]: We know that when we add FileIo here, that we won't need to do any manual flushing. Therefore, we only add the `is_dirty` flag on the MmapOnly variant.

**Comment by HaoranYi** [comment]
Add a counter for dirty mmap

**Comment by brooksprumo** [comment]
> Add a counter for dirty mmap

Done in https://github.com/anza-xyz/agave/pull/1676/commits/0782d345eb1940d49708fe419e41753d6d3bddf3.

**Review by jeffwashington — APPROVED** [review]
i like it

**Review by HaoranYi — APPROVED** [review]
Thanks! lgtm!

**Pull request merged**